### ISSUE TITLE
[HOLD for payment 2024-11-01] Expensify Card - Expensify Card link opens editable details page when card is deactivated

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#50375]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.46-1
**Reproducible in staging?:** Y
**Reproducible in production?:** N/A - new feature, doesn't exist in prod
**If this was caught during regression testing, add the test name, ID and link from TestRail:** applausetester+khcategory16@applause.expensifail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. [Admin] Asssign a virtual card to member.
3. [Admin] Go to Expensify Card.
4. [Admin] Click on the member with the virtual card in Step 2.
5. [Admin] Click Deactivate card.
6. [Admin] Deactivate the card.
7. [Admin] Go to workspace chat with the member in Step 2.
8. [Admin] Click on Expensify Card link.

## Expected Result:
Not here page should show up.

## Actual Result:
Expensify Card details page opens. Admin can edit the fields even though the card is already deactivated.

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/36b55038-43f1-4edf-9185-e86ae9bd6e65

 
</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Issue Owner</summary>Current Issue Owner: @johncschuster</details>

**Comment by IuliiaHerets** [comment]
We think that this bug might be related to #wave-collect - Release 2

**Comment by mountiny** [comment]
I will take over this one

**Comment by DylanDylann** [comment]
@truph01 This bug happens after your [PR](https://github.com/Expensify/App/pull/50039) is merged. Could you please take a look at this issue?

**Comment by jjcoffee** [comment]
Hmm I don't think this is really a regression from https://github.com/Expensify/App/pull/50039. That PR just allows easier access to this route that was already missing access checks as in this [proposal](https://github.com/Expensify/App/issues/50375#issuecomment-2399314154). So on production if you copy the link before deactivating the card and open it after you also can access the details.

I also don't think this is really a blocker, as it's not a particularly big deal if you can still edit the card details.

**Comment by mountiny** [comment]
I think the solution here is that on the Card Details page, we need to look if the cardID from the route is available, if the card was deactivated, it was removed from onyx so it would not be there

@jjcoffee I think we need to treat it as a regression from that PR, because it enables this flow to be much more exposed

@daledah I believe this problem is for both pages - for admin and the wallet page for the user

**Comment by truph01** [comment]
@mountiny @DylanDylann I think this bug is not related to my PR, since it can be reproduced in production:

https://github.com/user-attachments/assets/cfd36b05-afae-443d-8670-91fad2b2fa74

**Comment by mountiny** [comment]
@truph01 you are deeplinking, which is not a normal user flow.

After your PR, there is a link that takes you to this page, so I consider it as a consequence of the PR. Are you able to look into a fix please?

**Comment by truph01** [comment]
@mountiny @daledah Sure, I am investigating it now

**Comment by truph01** [comment]
@jjcoffee 
- To address the issue, I plan to update this:

https://github.com/Expensify/App/blob/41c0b9ddd5576cf216db63cfc0d40ef74995fadc/src/pages/workspace/expensifyCard/WorkspaceExpensifyCardDetailsPage.tsx#L74-L75
to:

````javascript
    const isLoadingCard = isLoadingCardListFromOnyx || card?.isLoading;
    const shouldDisplayNotFoundPage = !isLoadingCard && !card?.cardID;

    if (isLoadingCard) {
        return <FullScreenLoadingIndicator />;
    }

    return (
        <AccessOrNotFoundWrapper
            shouldBeBlocked={shouldDisplayNotFoundPage}
````

- The ```card?.isLoading === false``` is to check whether the card data is fetched successfully via API call:

https://github.com/Expensify/App/blob/41c0b9ddd5576cf216db63cfc0d40ef74995fadc/src/pages/workspace/expensifyCard/WorkspaceExpensifyCardDetailsPage.tsx#L60-L62

To set the ```isLoading``` value, this [function](https://github.com/Expensify/App/blob/41c0b9ddd5576cf216db63cfc0d40ef74995fadc/src/libs/actions/Card.ts#L647) should be updated to:

````javascript

function openCardDetailsPage(cardID: number, workspaceAccountID?: string) {
    const authToken = NetworkStore.getAuthToken();
    const optimisticData: OnyxUpdate[] = [];
    const successData: OnyxUpdate[] = [];
    const failureData: OnyxUpdate[] = [];
    if (!authToken) {
        return;
    }
    if (workspaceAccountID) {
        optimisticData.push({
            onyxMethod: Onyx.METHOD.MERGE,
            key: `${ONYXKEYS.COLLECTION.WORKSPACE_CARDS_LIST}${workspaceAccountID}_${CONST.EXPENSIFY_CARD.BANK}`,
            value: {
                [cardID]: {
                    isLoading: true,
                },
            },
        });

        successData.push({
            onyxMethod: Onyx.METHOD.MERGE,
            key: `${ONYXKEYS.COLLECTION.WORKSPACE_CARDS_LIST}${workspaceAccountID}_${CONST.EXPENSIFY_CARD.BANK}`,
            value: {
                [cardID]: {
                    isLoading: false,
                },
            },
        });

        failureData.push({
            onyxMethod: Onyx.METHOD.MERGE,
            key: `${ONYXKEYS.COLLECTION.WORKSPACE_CARDS_LIST}${workspaceAccountID}_${CONST.EXPENSIFY_CARD.BANK}`,
            value: {
                [cardID]: {
                    isLoading: false,
                },
            },
        });
    }

    const parameters: OpenCardDetailsPageParams = {
        authToken,
        cardID,
    };

    API.read(READ_COMMANDS.OPEN_CARD_DETAILS_PAGE, parameters, {optimisticData, successData, failureData});
}
````

and update:
https://github.com/Expensify/App/blob/41c0b9ddd5576cf216db63cfc0d40ef74995fadc/src/pages/workspace/expensifyCard/WorkspaceExpensifyCardDetailsPage.tsx#L60-L62
to:
````javascript
    const fetchCardDetails = useCallback(() => {
        if (!cardID || !workspaceAccountID) {
            return;
        }

        Card.openCardDetailsPage(Number(cardID), String(workspaceAccountID));
    }, [cardID, workspaceAccountID]);
````
- All the changes work well:


https://github.com/user-attachments/assets/c042fc76-c0e4-4ce8-896b-10eb14c73527




Do you have any feedback about this approach?

**Comment by mountiny** [comment]
I think this is good for now, but we should think through a better UX

**Comment by truph01** [comment]
> you are deeplinking, which is not a normal user flow.
 
Using a deep link is simply the easiest way to reproduce the bug. Alternatively, you can reproduce the bug in production with the following steps::

1. Open the card detail page in two tabs.

2. Deactivate the card in tab 2.

3. There is no "not here page" in tab 1.


https://github.com/user-attachments/assets/059624a9-aef4-4f05-a5bd-1fbc6a71d24d

**Comment by mountiny** [comment]
@truph01 I am not disagreeing with you, I am just noting that normal user would not run into this. People are not copying cardIDs and then composing the urls based on that.

**Comment by truph01** [comment]
@jjcoffee Please take a look at my comment [above](https://github.com/Expensify/App/issues/50375#issuecomment-2399546386), then I open PR.

**Comment by mountiny** [comment]
@truph01 can you please raise the PR?

**Comment by truph01** [comment]
I am implementing PR

**Comment by truph01** [comment]
@mountiny After applying my changes in [PR](https://github.com/Expensify/App/pull/50404), I encountered a few BE issues that blocked my testing. Here are the steps:

1. In device A, login as user A (member). In device B, login as user B (admin).
2. [User B]: Issue a new card to user A.
3. [User A]: Open the workspace chat. Observe that there is no "issued @truphd01+1001@gmail.com a virtual Expensify Card! The card can be used right away." message (**Bug 1**)
4. [User A]: Fresh the page. The message mentioned in step 3 appears.
5. [User A]: Click on the Expensify Card link in the message mentioned in step 3.
6. [User A]: Verify the Not Found page appears. (**Bug 2**)
7. [User A]: Clear cache and restart.
8. [User A]: Click on the Expensify Card link in the message mentioned in step 3.
9. [User A]: Verify the card detail is displayed properly.

Do you think we need to fix these bugs from BE side first?

**Comment by truph01** [comment]
> Can you share video for bug 2 @truph01 ?

@allgandalf I updated the video for two bugs in [above comment](https://github.com/Expensify/App/issues/50375#issuecomment-2399834159)

**Comment by truph01** [comment]
> the not found page will get triggered if a certain thing is missing right

It is because the ```cardsList``` data does not contain the data of the target cardID.

**Comment by mountiny** [comment]
@truph01 we do push these changes using pusher so it just might be that there are some delays, I would not worry about it now as long as you can fix the bug in this issue

**Comment by thienlnam** [comment]
Had a discussion [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1728413452162699), but TLDR is 

Marking this as a non-blocker since it’s a flow that I don’t anticipate many people will run into on NewDot yet, and then we can follow up with a more complete fix once Vit and Joel are back online

**Comment by johncschuster** [comment]
Bumping for Melv. Looks like we're just waiting on @mountiny and @jjcoffee to come back

**Comment by mountiny** [comment]
I think we are going to add a new report action when the card is deactivated, with that, I think we can just remove the link from the message when the card is not locally available in the `cardList`

**Comment by mountiny** [comment]
@truph01 @jjcoffee could you please update the solution so the report action **does not** have a link in case the cardID is not locally stored in the cardList or in the cards_ (in case of admin)

**Comment by truph01** [comment]
I am investigating

**Comment by truph01** [comment]
> could you please update the solution so the report action does not have a link in case the cardID is not locally stored in the cardList or in the cards_ (in case of admin)

@jjcoffee Can you pre-review my solution in this draft [PR](https://github.com/Expensify/App/pull/50720)? Then I will add all videos then open PR.

**Comment by jjcoffee** [comment]
@truph01 The approach looks good, please proceed!

**Comment by truph01** [comment]
[PR](https://github.com/Expensify/App/pull/50720) is ready

**Comment by jjcoffee** [comment]
I don't think any payment is due here as this was a regression fix.

**Comment by johncschuster** [comment]
Thanks, @jjcoffee! 👍

**Comment by johncschuster** [comment]
I think we'll still need the BZ Checklist completed. Can you complete that when you get a moment?

**Comment by jjcoffee** [comment]
- The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/50039
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/50039#issuecomment-2456572296
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A - more thorough testing was required
- Determine if we should create a regression test for this bug. **Yes**
- If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

**Regression Test Proposal**

1. [Admin] Asssign a virtual card to member.
2. [Admin] Go to Expensify Card.
3. [Admin] Click on the member with the virtual card in Step 2.
4. [Admin] Click Deactivate card and deactivate the card.
5. [Admin] Go to workspace chat with the member in Step 2.
6. [Admin] Verify in the issue card message that the `Expensify Card` text is not a link.

Do we agree 👍 or 👎
